### PR TITLE
feat(HACBS-1728): implement appropriate metrics for the Internal Service controller

### PR DESCRIPTION
**Pull request by theflockers**
feat(HACBS-1728): implement metrics for the Internal Service controller
    
- adds metrics for internal-services-controller
- update go.mod
- adds tests for /metrics
    
Signed-off-by: Leandro Mendes <lmendes@redhat.com>

### CODE REVIEW

File: controllers/internalrequest/adapter_test.go [hunk 1]
```diff
@@ -211,6 +212,8 @@ var _ = Describe("PipelineRun", Ordered, func() {
 		})
 
 		It("should mark the InternalRequest as running", func() {
+			adapter.internalRequest.Status.StartTime = nil
```
Comment (davidmogar) [inline controllers/internalrequest/adapter_test.go:215]: Why is this here? The test was working before.

File: metrics/internalrequest.go [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+		prometheus.GaugeOpts{
+			Name: "internal_request_attempt_concurrent_requests",
+			Help: "Total number of concurrent InternalRequest attempts",
+		},
+	)
+
+	InternalRequestAttemptDurationSeconds = prometheus.NewHistogramVec(
+		prometheus.HistogramOpts{
+			Name:    "internal_request_attempt_duration_seconds",
+			Help:    "InternalRequest durations from the moment the InternalRequest PipelineRun was created til the InternalRequest is marked as finished",
```
Comment (davidmogar) [inline metrics/internalrequest.go:38]: Let's reword it. I know you are copying what I did in the release-service, but I'm the wrong person to copy from :stuck_out_tongue: 
```suggestion
			Help:    "Time from the moment the InternalRequest starts being processed til it completes",
```

File: metrics/internalrequest.go [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+			Buckets: []float64{60, 150, 300, 450, 600, 750, 900, 1050, 1200, 1800, 3600},
+		},
+		[]string{"request", "namespace", "reason", "succeeded"},
+	)
+
+	InternalRequestAttemptRunningSeconds = prometheus.NewHistogram(
```
Comment (davidmogar) [inline metrics/internalrequest.go:44]: We don't need this one. Not much validation happening here. Not interesting at all.

File: metrics/internalrequest.go [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+			Name: "internal_request_attempt_concurrent_requests",
+			Help: "Total number of concurrent InternalRequest attempts",
+		},
+	)
+
+	InternalRequestAttemptDurationSeconds = prometheus.NewHistogramVec(
+		prometheus.HistogramOpts{
+			Name:    "internal_request_attempt_duration_seconds",
+			Help:    "InternalRequest durations from the moment the InternalRequest PipelineRun was created til the InternalRequest is marked as finished",
+			Buckets: []float64{60, 150, 300, 450, 600, 750, 900, 1050, 1200, 1800, 3600},
```
Comment (davidmogar) [inline metrics/internalrequest.go:39]: These numbers make sense in the context of the release-service, but are we expecting 1 hour long requests? Don't think so.

**Review by davidmogar — CHANGES_REQUESTED** [review]
Partial review as changes are requested.

File: controllers/internalrequest/adapter_test.go (hunk 1 above)
Comment (theflockers) [reply controllers/internalrequest/adapter_test.go:215]: Because I added `MarkRunning()`  in the line 320. Otherwise it breaks because StartTime is not set (as we added the `go.metrics` to `MarkFailed()`and `MarkSucceeded()`.

File: metrics/internalrequest.go (hunk 4 above)
Comment (theflockers) [reply metrics/internalrequest.go:39]: Good question. So far I noticed an average of 5 minutes processing on IIB side. I will ask them the average build time and worst case scenario we should expect on their side.

File: controllers/internalrequest/adapter_test.go (hunk 1 above)
Comment (davidmogar) [reply controllers/internalrequest/adapter_test.go:215]: Yeah... about that `MarkRunning` call, it should be there. Not every test needs that function to get called, so it's better to call it for tests that need it.

File: metrics/internalrequest.go (hunk 4 above)
Comment (davidmogar) [reply metrics/internalrequest.go:39]: This is not only for IIB, but I don't think 1 hour request is that normal... Maybe it is though. @scoheb Can you shred some light here? What internal services could we be reaching that have a long time to run? Brew came to my head, but we won't use it.

Comment (theflockers) [reply metrics/internalrequest.go:39]: @davidmogar I discovered last week that a request will time out after 1h on IIB  - and Yes, it took that long due to a big queue on their side. So I think we are safe to keep the numbers.

File: metrics/internalrequest.go (hunk 3 above)
Comment (theflockers) [reply metrics/internalrequest.go:44]: @davidmogar correct of I am wrong: This one means the time the internalrequest is created, the Pod is created, bootstraped and then running, hence, the Pipeline is marked as running. Maybe would it be an interesting clue that could help to identify a potential problem?

Comment (davidmogar) [reply metrics/internalrequest.go:44]: In the release service this is used to see how much we take to validate a Release. In the Internal Services operator, this would be the time until we create the PipelineRun (not running). So that's probably less than a second.

File: metrics/internalrequest.go (hunk 4 above)
Comment (davidmogar) [reply metrics/internalrequest.go:39]: I still don't think so. Most of the requests will take less than a minute, which is the minimum value/bucket. Better values, in my opinion, would be something like `10, 20, 40, 60, 150, 300, 450, 900, 1800, 3600`. With those we get more value from the first buckets.

Comment (theflockers) [reply metrics/internalrequest.go:39]: IIB operations takes at least 5 minutes to finish.

Comment (davidmogar) [reply metrics/internalrequest.go:39]: This is not an IIB operator. Also, one particular IIB operation takes that time to finish. I doubt that IIB get endpoints take, all of them, that much. Feel free to adjust the buckets, but we don't need precision for requests longer than let's say 10 minutes.

Comment (theflockers) [reply metrics/internalrequest.go:39]: I think the numbers you suggested look better. And Indeed, probably other endpoints other than IIB might behave very different.

*Just for the record* - I would say that this IIB operation in particular (we were using to test) is one of the fastest possible build as it contains only a single FBC fragment. This could be considerably worse depending on the number of components it has to build. But yes, this is a generic, not an IIB operator.

File: metrics/internalrequest_test.go [hunk 5]
```diff
@@ -0,0 +1,122 @@
...
+var _ = Describe("Metrics InternalRequest", Ordered, func() {
+	BeforeAll(func() {
+
+		// We need to unregister in advance otherwise it breaks with 'AlreadyRegisteredError'
+		metrics.Registry.Unregister(InternalRequestAttemptConcurrentTotal)
+		metrics.Registry.Unregister(InternalRequestAttemptDurationSeconds)
+	})
+
+	var (
+		AttemptDurationSecondsHeader = inputHeader{
```
Comment (davidmogar) [inline metrics/internalrequest_test.go:43]: I know they are like this in the release-service, but why are they capitalized? They are only used in this package.

Comment (theflockers) [reply metrics/internalrequest_test.go:43]: I have no idea why. Maybe there was a reason back then. I will correct it here, at least.

File: metrics/internalrequest.go [hunk 6]
```diff
@@ -0,0 +1,81 @@
...
+		"reason":    reason,
+		"succeeded": strconv.FormatBool(succeeded),
+	}
+	InternalRequestAttemptConcurrentTotal.Dec()
+	InternalRequestAttemptDurationSeconds.With(labels).Observe(completionTime.Sub(startTime.Time).Seconds())
+	InternalRequestAttemptTotal.With(labels).Inc()
+}
+
+// RegisterNewInternalRequest increments the 'internal_request_attempt_concurrent_total' and registers a new observation for
+// 'internal_request_attempt_duration_seconds' with the elapsed time from the moment the InternalRequest was created to when
```
Comment (davidmogar) [inline metrics/internalrequest.go:69]: It only changes one metric.

File: metrics/internalrequest_test.go [hunk 7]
```diff
@@ -0,0 +1,122 @@
...
+		}
+	)
+
+	const (
+		validInternalRequestReason   = "valid_internalrequest_reason"
+		invalidInternalRequestReason = "invalid_internalrequest_reason"
+		strategy                     = "nostrategy"
+	)
+
+	var defaultNamespace = "default"
```
[the remark this answers is not in the document]
Comment (theflockers) [reply metrics/internalrequest_test.go]: moved.

File: metrics/internalrequest.go (hunk 6 above)
Comment (theflockers) [reply metrics/internalrequest.go]: changed. PTAL.

**Review by davidmogar — APPROVED** [review]

**Review by davidmogar — APPROVED** [review]
